src/librustc/dep_graph/graph.rs

```diff
@@ -758,71 +759,89 @@ impl DepGraph {
 
         // ... emitting any stored diagnostic ...
 
+        // FIXME: Store the fact that a node has diagnostics in a bit in the dep graph somewhere
+        // Maybe store a list on disk and encode this fact in the DepNodeState
         let diagnostics = tcx.queries.on_disk_cache
-                                .load_diagnostics(tcx, prev_dep_node_index);
+                             .load_diagnostics(tcx, prev_dep_node_index);
+
+        #[cfg(not(parallel_compiler))]
+        debug_assert!(data.colors.get(prev_dep_node_index).is_none(),
+                      "DepGraph::try_mark_previous_green() - Duplicate DepNodeColor \
+                      insertion for {:?}", dep_node);
 
         if unlikely!(diagnostics.len() > 0) {
             self.emit_diagnostics(
                 tcx,
                 data,
                 dep_node_index,
-                did_allocation,
+                prev_dep_node_index,
                 diagnostics
             );
+        } else {
+            // Avoid calling the destructor, since LLVM fails to optimize it away
+            mem::forget(diagnostics);
```

Comment:
`diagnostics` is just an empty `Vec` here, with the current types, so I agree that a `mem::forget` here should have the same semantics as running the destructor.

But I'm finding it very surprising that the destructor would be that much of a burden. Did you actually find it to be an issue in execution profiles?

I just am trying to understand whether this `else { ... }` block is paying for itself, in the long term maintenance costs for future developers having to re-prove in their heads that this is not leaking memory.

Reply:
I have another optimization in mind which would deal with it, so I could just remove it. I don't know how it affects runtime, but removing code can only improve it.

Reply:
I removed it.